src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IsPatternOperator.cs

```diff
@@ -14,310 +17,120 @@ internal sealed partial class LocalRewriter
-                            foreach (BoundPatternBinding binding in whenNode.Bindings)
+                        case BoundTestDecisionDagNode testNode:
+                            if (testNode.WhenTrue is BoundEvaluationDecisionDagNode e &&
+                                TryLowerTypeTestAndCast(testNode.Test, e.Evaluation, out BoundExpression? sideEffect, out BoundExpression? testExpression))
                             {
-                                BoundExpression left = _localRewriter.VisitExpression(binding.VariableAccess);
-                                BoundExpression right = _tempAllocator.GetTemp(binding.TempContainingValue);
-                                if (left != right)
-                                {
-                                    _sideEffectBuilder.Add(_factory.AssignmentExpression(left, right));
-                                }
+                                return _factory.LogicalAnd(_factory.Sequence(ImmutableArray<LocalSymbol>.Empty, ImmutableArray.Create(sideEffect), testExpression), map[e.Next]);
                             }
-                        }
-
-                        break;
-
-                    default:
-                        throw ExceptionUtilities.UnexpectedValue(node.Kind);
-                }
-
-                if (_sideEffectBuilder.Count > 0 || _conjunctBuilder.Count == 0)
-                {
-                    AddConjunct(_factory.Literal(true));
-                }
-
-                Debug.Assert(_sideEffectBuilder.Count == 0);
-                BoundExpression? result = null;
-                foreach (BoundExpression conjunct in _conjunctBuilder)
-                {
-                    result = (result == null) ? conjunct : _factory.LogicalAnd(result, conjunct);
-                }
 
-                _conjunctBuilder.Clear();
-                Debug.Assert(result != null);
-                var allTemps = _tempAllocator.AllTemps();
-                if (allTemps.Length > 0)
-                {
-                    result = _factory.Sequence(allTemps, ImmutableArray<BoundExpression>.Empty, result);
+                            BoundExpression whenTrue = map[testNode.WhenTrue];
+                            BoundExpression whenFalse = map[testNode.WhenFalse];
+                            return (whenTrue.ConstantValue?.BooleanValue, whenFalse.ConstantValue?.BooleanValue) switch
+                            {
+                                (true, true) => whenTrue,
+                                (false, false) => whenFalse,
+                                (true, false) => LowerTest(testNode.Test),
+                                (false, true) => _factory.Not(LowerTest(testNode.Test)),
+                                (null, true) => _factory.LogicalOr(_factory.Not(LowerTest(testNode.Test)), whenTrue),
+                                (null, false) => _factory.LogicalAnd(LowerTest(testNode.Test), whenTrue),
```

Comment:
Couple of issues with this approach:
1. This expression is deep on the right-hand-side and emitter generates a different (worse) code for that.
2. There a *chance* to produce duplicated checks because the entire graph is inlined. (no test coverage there)
3. This should be able to incorporate switch optimizations: https://github.com/dotnet/roslyn/issues/45679